test/common/bidi.py

```diff
@@ -0,0 +1,438 @@
+import sys
+import tempfile
+from dataclasses import dataclass
+from pathlib import Path
+from typing import Any
+
+import aiohttp
+
+log_proto = logging.getLogger("bidi.proto")
+log_command = logging.getLogger("bidi.command")
+
+
+class WebdriverError(RuntimeError):
+    pass
+
+
+@dataclass
+class LogMessage:
+    level: str  # like "info"
+    type: str  # usually "console"
+    timestamp: int
+    args: list[object]
+    text: str
+
+    def __init__(self, message_params):
+        self.level = message_params["level"]
+        self.type = message_params["type"]
+        self.timestamp = message_params["timestamp"]
+        self.args = message_params.get("args", [])
+        self.text = message_params["text"]
+
+    def __str__(self):
+        # mostly for naughty compatibility with CDP
+        level = self.level.replace("warn", "warning")
+        return f"> {level}: {self.text}"
+
+
+@dataclass
+class BidiSession:
+    ws_url: str
+    session_url: str
+    process: asyncio.subprocess.Process
+
+
+# Return port numbers that were free at the time of checking
+# They might be in use again by the time the function returns...
+def pick_ports(count: int) -> list[int]:
+    sockets: list[socket.socket] = []
+    ports: list[int] = []
+
+    for _ in range(count):
+        sock = socket.socket()
+        sock.bind(('127.0.0.1', 0))
+        sockets.append(sock)
+        ports.append(sock.getsockname()[1])
+
+    for s in sockets:
+        s.close()
+
+    return ports
+
+
+def jsquote(js: object) -> str:
+    return json.dumps(js)
+
+
+def unpack_value(raw: Any) -> Any:
+    """Convert a WebDriver type/value object into a plain Python object"""
+
+    if not isinstance(raw, dict):
+        return raw
+    try:
+        _type = raw["type"]
+    except KeyError as e:
+        raise ValueError(f"No type in {raw!r}") from e
+    if _type == "undefined":
+        return None
+    if _type in ["string", "boolean", "number"]:
+        return raw["value"]
+    # serialized BiDi objects, don't touch
+    if _type in ["function", "node", "window"]:
+        return raw
+    if _type == "array":
+        return [unpack_value(v) for v in raw["value"]]
+    if _type == "object":
+        obj = {}
+        for k, v in raw["value"]:
+            obj[k] = unpack_value(v["value"])
+        return obj
+    raise ValueError(f"Unknown type in {raw!r}")
+
+
+class WebdriverBidi:
+    http_session: aiohttp.ClientSession
+
+    def __init__(self, headless: bool = False) -> None:
+        self.headless = headless
+        self.last_id = 0
+        self.pending_commands: dict[int, asyncio.Future] = {}
+        self.logs: list[LogMessage] = []
+        self.bidi_session: BidiSession | None = None
+        self.future_wait_page_load = None
+        self.top_context: str | None = None  # top-level browsingContext
+        self.context: str | None  # currently selected context (top or iframe)
+
+        self.homedir = tempfile.TemporaryDirectory(prefix="browser-home-")
```

Comment:
I really *really* don't want to introduce the mess of making `WebdriverBidi` a context manager, we already discussed that in the starter-kit demo. This totally doesn't fit into how it is being used in Browser, and would make this awkward/unwieldy to use.

`TemporaryDirectory` *can* be used as a context manager, but doesn't have to be. The TempDir object is destructed together with the WebdriverBidi object, it doesn't leave the directory behind.

Added the "cockpit-test-*" prefix and changed so that `self.homedir` is a Path.

Reply:
Relying on this implicit destruction via garbage collection is unpythonic.  Something should definitely be explicitly closing this.

Reply:
Honest question: citation? The whole point of TemporaryDirectory over mkdtemp was that it automatically cleans up after itself.. I could add `self.homedir_temp.cleanup()` to close(), but then this isn't conceptually right either -- accessing the value after .close() Should Not Happen™, but in the general case it might. And I can't set it to `= None` as that's an incompatible type. And I'll surely not set it to `TempDir | None` as that'd require an assert at every caller. This is just making life unnecessarily hard AFAICS..